### ISSUE TITLE
predicted probabilities are almost the same for each class

### ISSUE THREAD

**Issue by geoHeil** [Microsoft/LightGBM#272]
When calling `lgb.predict_proba(X_new)` I get the following result. As you can see, there are almost the same probabilities for the histogram.
![lightgbmexploration](https://cloud.githubusercontent.com/assets/1694964/22423110/337fa114-e6f0-11e6-90f2-e02fbe1082b8.jpg)

When setting `raw_score=TRUE` do I understand correctly that this would give me the values before the logit transformation?

![download](https://cloud.githubusercontent.com/assets/1694964/22423567/851b253c-e6f2-11e6-9457-8ff4783a621a.png)

But what is the influence of num_iterations in the predict_proba stage when RAW=True?
With 1 iterations:
![1](https://cloud.githubusercontent.com/assets/1694964/22423632/d367ef0e-e6f2-11e6-93c9-3b546a6ba0bd.png)

and with 200
![200](https://cloud.githubusercontent.com/assets/1694964/22423638/d8ed2368-e6f2-11e6-9eb3-e5d0b56bdf13.png)

**Comment by guolinke** [comment]
I think it will output probabilities, not 0/1 if using predict_prob.

@wxchan  do you have any idea?

**Comment by Laurae2** [comment]
> As you can see, there are almost the same probabilities for the histogram.

Without knowing your data, your output seems normal. The model is apparently very good at separating classes 0 and 1 with only one tree (the predicted values are not 0 and 1, a sigmoid can't predict 0 and 1 unless the values are negatively/positively infinite), and therefore with 200 trees it attempts to calibrate the probabilities to minimize the logarithmic loss (if you use binary classification defaults).

If you want to discern probabilities more clearly, I suggest you take a dataset where it is not obvious to separate the classes. Example: https://www.kaggle.com/c/bnp-paribas-cardif-claims-management/data - this one you can not separate classes nearly perfectly, unlike your dataset where it is near perfect separation.

> When setting raw_score=TRUE do I understand correctly that this would give me the values before the logit transformation?

Yes. If you compute the sigmoid `1/(1+exp(-x))` from the raw scores (take for example `-10, -5, 5, 10`), you will see the probabilities are both on the edges of 0 and 1, they are not identical. Your raw score histogram associates with your probability picture, there is no issue (you can still output us the table of unique values, they cannot be only 0 and 1 due to the sigmoid application - if they are really 0 and 1, it is an issue to fix).

> But what is the influence of num_iterations in the predict_proba stage when RAW=True?

num_iterations parameter changes the number of boosted trees used for the predictions. For instance, if you trained for 1000 iterations and you use `num_iterations = 500`, it is the same as training for 500 iterations and predicting from that model.

**Comment by geoHeil** [comment]
@Laurae2 thanks for that input. Indeed values are around 0.9 and not 1 respectively the same for 0.1 and not 0.

Why would I train for 1000 and evaluate only on 500? Isn't tree based pruning more efficient here or a max_depth limit?

**Comment by Laurae2** [comment]
> Why would I train for 1000 and evaluate only on 500? Isn't tree based pruning more efficient here or a max_depth limit?

@geoHeil boosting is an ensemble prone to overfitting which must be controlled using a separate validation dataset. For instance, if you know from the validation evaluation metric you are overfitting past 500 trees (whatever your other parameters are, it does not matter), you would rather avoid predicting more than 500 trees to have the optimal model according to the validation evaluation metric.

Example:

| Boosting iteration | Evaluation logloss |
| --: | --: |
| 1 | 0.5200 |
| 10 | 0.2100 |
| 50 | 0.1900 |
| 100 | 0.1750 |
| 200 | 0.1720 |
| 499 | 0.1965 |
| **500** | **0.1694** |
| 501 | 0.1695 |
| 502 | 0.1698 |
| 1000 | 0.4950 |

You would predict here for 500 trees according to the best validation evaluation metric result.

When you use boosting, you usually use an early stopping method (`early_stopping_rounds` for instance in xgboost / LightGBM) to automatically stop the training when the validation score does not get better after X rounds. You also usually put an infinite amount (or any large acceptable amount depending on the time you want to spend training) of training iterations so your training is not interrupted early because you didn't allow enough iterations.

Another usage of training for a specific amount of iterations without early stopping is for debugging purposes (like: showing a training graph to compare with another model, show evaluation of metric and the overfitting, etc.).

In practice the number of iterations is not an hyperparameter to tune because the validation method should provide you the optimal number of iterations. It is more a constraint of the time you can spend training a model (it must be adjusted with the learning rate - a lower learning rate means inversely more iterations needed to achieve similar results - so the time spent training a model can get huge quickly!).

**Comment by geoHeil** [comment]
Thanks for this great writeup.
